### ISSUE TITLE
Strange default number of slots in v2.0.x with hostfile

### ISSUE THREAD

**Issue by jsquyres** [open-mpi/ompi#2565]
I accidentally ran with an odd configuration today at the v2.0.x head, but the slot count was not what I expected it to me.

# Case 1

Ran with:

```
$ cat hostfile
# each host is a sandy bridge, 16 cores
host-sandy1
host-sandy2
host-sandy3
host-sandy4
$ mpirun --hostfile hostfile hostname | wc -l
64
```

That's expected.

# Case 2

In addition to the 4 x sandy bridges, I ran with a 20-core ivy bridge:

```
$ cat hostfile
# same hosts as above, but with an additional 20-core ivy bridge
host-sandy1
host-ivy1
host-sandy2
host-sandy3
host-sandy4
$ mpirun --hostfile hostfile hostname | wc -l
112
```

112?  I expected 64 + 20 = 84.  A little more detail:

```
$ mpirun --hostfile hostfile hostname | sort | uniq -c
     16 host-sandy1
     24 host-ivy1
     24 host-sandy2
     24 host-sandy3
     24 host-sandy4
```

This was a surprise to me.  I expected 16 on each of the sandys and 20 on the ivy (although I didn't actually intend to run with a mix of ivy and sandys, this is what happened).

Is this expected behavior?

**Comment by rhc54** [comment]
Hard to say what might happen - if the nodes aren't the same, you are supposed to add ```--hetero-nodes``` to the mpirun cmd line

**Comment by jsquyres** [comment]
I think you're saying "if you're in a hetero situation and don't tell us, the behavior is undefined."  Right?  If so, that sounds like a fair point to me.

It would be nice if we could somehow figure out and warn (or abort?) if we're in a hetero situation and `--hetero-nodes` was not used (even if it's after the mapping/binding is done -- e.g., if an `orted` realizes, "hey, I don't have the same topology as what I received in the launch message from the HNP" and then warns / aborts.  I.e., it would be *late* warning/abort (vs. detecting it at mpirun / the HNP), but it would still do so.

Is that possible?

**Comment by jsquyres** [comment]
@rhc54 and I talked about this on the phone.

I tested on master and again on v2.0.x head this morning.

Running with `--hetero-nodes` *does* seem to do the Right Thing (i.e., get 16 on each sandy and 20 on ivy).

However, this morning, even running *without* `--hetero-nodes`, I *also* get the Right Thing.  ☹️ 

@rhc54 Should I attribute that to undefined behavior when heterogeneous nodes are mixed together -- sometimes it'll do the right thing, but other times it won't?  If so, how can we prove that `--hetero-nodes` works?

**Comment by rhc54** [comment]
I'm honestly not sure - it probably depends a lot on where mpirun lands and where the first daemon lands. Hit things right, and you might get lucky.

As I said on the phone, I have code in master that was intended to help identify this situation, but I'm not sure the "signature" is good enough to differentiate such similar machines. Probably need to come up with something else.

**Comment by rhc54** [comment]
Actually, I thought about this some more and realized that everything is working as it should. mpirun generates a "signature" string that includes the number of each type of HWLOC object. So it tells the receiver "I have x NUMA, y sockets, z cores, etc.". Each orted gets that info on its cmd line, and then checks it against their own local "signature". If the two differ, then we assume that node is different than the one mpirun is on, and we send back the topology.

All ```--hetero-nodes``` does is force the orted to send its local topology regardless of the signature string. The option is necessary for those cases where all the number of object types are the same, but the relationship (or hierarchy) between them is different. I don't know how likely that scenario is, but couldn't completely discard it.

So everything on master at least should be working correctly, and I believe that the code exists on the 2.0/2.x branches as well, at least in terms of detection. It sounds, however, like the slot count isn't being correctly computed (or the linkage of node to topology may be off), so that might merit some further investigation.

**Comment by rhc54** [comment]
I don't see anything wrong.  Could you please rerun your tests with ```-mca plm_base_verbose 5``` on master and send me the output when it _doesn't_ do the right thing?

**Comment by hppritcha** [comment]
Where are we on this?

**Comment by hppritcha** [comment]
If something needs to be fixed we will do so in 2.0.3.

**Comment by rhc54** [comment]
We now have auto-detect working in master - it can be brought over to 2.0.3, I suppose, if we really want. Otherwise, I would schedule it for 2.1.

Marking this as fixed.
